src/backend/distributed/utils/metadata_cache.c

```diff
@@ -2912,6 +2934,51 @@ InvalidateDistRelationCacheCallback(Datum argument, Oid relationId)
 }
 
 
+/*
+ * SetRelatedOidsInvalid sets the affected relation's cache entry to invalid.
+ * It also invalidates foreign key relation graph if any relation
+ */
+static void
+SetRelatedOidsInvalid(Oid relationId)
+{
+	List *connectedComponentListOfRelation = NIL;
+	ListCell *nodeCell = NULL;
+	bool foundInCache = false;
+
+	/*
+	 * We need to invalidate cache entry of each relation which are in the
+	 * same connected component with the given relation. If we do not have
+	 * foreign key relation graph, we don't need to do anything.
+	 *
+	 * We also need to invalidate foreign key relation graph if any of the
+	 * node in it is affected by invalidating given relation ID.
+	 */
+	if (IsForeignKeyGraphValid())
```

Comment:
Isn't it possible to get two invalidation messages, and since you invalidate the whole graph on the first message (and don't rebuild it), you'll end up not processing the second invalidation?